// lichess-pgn-viewer/dist/util.js
export function bindMobileMousedown(el, f, redraw) {
    for (const mousedownEvent of ['touchstart', 'mousedown']) {
        el.addEventListener(mousedownEvent, e => {
            f(e);
            e.preventDefault();
            if (redraw)
                redraw();
        }, { passive: false });
    }
}
export const bindNonPassive = (eventName, f, redraw) => bind(eventName, f, redraw, false);
export const bind = (eventName, f, redraw, passive = true) => onInsert(el => el.addEventListener(eventName, e => {
    const res = f(e);
    if (res === false)
        e.preventDefault();
    redraw === null || redraw === void 0 ? void 0 : redraw();
    return res;
}, { passive }));
export function onInsert(f) {
    return {
        insert: vnode => f(vnode.elm),
    };
}
export function stepwiseScroll(inner) {
    let scrollTotal = 0;
    return (e) => {
        scrollTotal += e.deltaY * (e.deltaMode ? 40 : 1);
        if (Math.abs(scrollTotal) >= 4) {
            inner(e, true);
            scrollTotal = 0;
        }
        else {
            inner(e, false);
        }
    };
}
export function eventRepeater(action, e) {
    const repeat = () => {
        action();
        delay = Math.max(100, delay - delay / 15);
        timeout = setTimeout(repeat, delay);
    };
    let delay = 350;
    let timeout = setTimeout(repeat, 500);
    action();
    const eventName = e.type == 'touchstart' ? 'touchend' : 'mouseup';
    document.addEventListener(eventName, () => clearTimeout(timeout), { once: true });
}
