// lichess-pgn-viewer/dist/view.js
import { Chessground } from 'chessground';
import { h } from 'snabbdom';
import { bind, bindMobileMousedown, bindNonPassive, eventRepeater, onInsert, stepwiseScroll } from './util';
export default function view(ctrl) {
    return ctrl.menu ? renderMenu(ctrl) : renderReplay(ctrl);
}
const renderMenu = (ctrl) => h('div.lpv.lpv--menu', [
    h('div.lpv__menu__inner', [
        h('button.lpv__menu__entry.lpv__menu__flip.lpv__fbt', {
            hook: bind('click', ctrl.flip),
        }, ctrl.translate('flipTheBoard')),
        h('a.lpv__menu__entry.lpv__menu__analysis.lpv__fbt', {
            attrs: {
                href: ctrl.analysisUrl(),
                target: '_blank',
            },
        }, ctrl.translate('analysisBoard')),
        h('a.lpv__menu__entry.lpv__menu__practice.lpv__fbt', {
            attrs: {
                href: ctrl.practiceUrl(),
                target: '_blank',
            },
        }, ctrl.translate('practiceWithComputer')),
    ]),
    renderControls(ctrl),
]);
const renderControls = (ctrl) => h('div.lpv__controls', [
    dirButton('backward', ctrl, -1),
    h('button.lpv__fbt.lpv__controls__menu', {
        class: { active: ctrl.menu },
        hook: bind('click', ctrl.toggleMenu),
    }, '⋮'),
    dirButton('forward', ctrl, 1),
]);
const dirButton = (name, ctrl, dir) => h(`button.lpv__controls__${name}.lpv__fbt`, {
    class: { disabled: !ctrl.menu && (ctrl.index + dir < 0 || ctrl.index + dir >= ctrl.nodes.length) },
    hook: onInsert(el => bindMobileMousedown(el, e => eventRepeater(() => ctrl.onward(dir), e))),
});
const renderReplay = (ctrl) => h('div.lpv.lpv--replay', [
    renderBoard(ctrl),
    renderControls(ctrl),
    ctrl.opts.showMoves ? renderMoves(ctrl) : undefined,
]);
const renderMoves = (ctrl) => h('div.lpv__moves', 'moves');
const renderBoard = (ctrl) => h('div.lpv__board', {
    hook: wheelScroll(ctrl),
}, h('div.cg-wrap', {
    hook: {
        insert: vnode => (ctrl.ground = Chessground(vnode.elm, makeConfig(ctrl))),
    },
}));
const wheelScroll = (ctrl) => 'ontouchstart' in window || !ctrl.opts.scrollToMove
    ? undefined
    : bindNonPassive('wheel', stepwiseScroll((e, scroll) => {
        e.preventDefault();
        if (e.deltaY > 0 && scroll)
            ctrl.onward(1);
        else if (e.deltaY < 0 && scroll)
            ctrl.onward(-1);
    }));
export const makeConfig = (ctrl) => ({
    viewOnly: true,
    addDimensionsCssVars: true,
    drawable: {
        enabled: false,
        visible: false,
    },
    ...ctrl.cgConfig(),
});
